File: week09/19/P5.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> tree;
        tree = kthSmallestHelp(root, tree);
        return tree[k-1];
    }

    vector<int> kthSmallestHelp(TreeNode* root, vector<int> tree) {
        if (root == nullptr) return tree;
        if (root->left != nullptr) tree = kthSmallestHelp(root->left, tree);
        tree.push_back(root->val);
        if (root->right != nullptr) tree = kthSmallestHelp(root->right, tree);
        return tree;
    }
};

int main() {
    TreeNode *root = new TreeNode(5);
    TreeNode *tree1 = new TreeNode(3);
    TreeNode *tree2 = new TreeNode(6);
    TreeNode *tree3 = new TreeNode(2);
    TreeNode *tree4 = new TreeNode(4);
    TreeNode *tree5 = new TreeNode(1);
    root->left = tree1;
    root->right = tree2;
    tree1->left = tree3;
    tree1->right = tree4;
    tree3->left = tree5;
    
    Solution result;
    cout << result.kthSmallest(root, 3) << endl;

    return 0;
}

File: week09/19/P5.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def kthSmallest(self, root, k):
        return self.kthSmallestHelp(root, [])[k-1]
    
    def kthSmallestHelp(self, root, tree):
        if root is None:
            return tree
        if root.left:
            tree = self.kthSmallestHelp(root.left, tree)
        tree.append(root.val)
        if root.right:
            tree = self.kthSmallestHelp(root.right, tree)
        return tree

if __name__ == "__main__":
    root = TreeNode(5, TreeNode(3, TreeNode(2, TreeNode(1)), TreeNode(4)), TreeNode(6))
    result = Solution()
    print(result.kthSmallest(root, 3))